### PR TITLE
[Updater] Add binary self update

### PR DESCRIPTION
**Pull request by Herklos**
Related to https://github.com/Drakkar-Software/OctoBot-Tentacles/pull/420

### CODE REVIEW

File: octobot/updater/updater_factory.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+import octobot_commons.os_util as os_util
+import octobot_commons.enums as commons_enums
+
+import octobot.updater.binary_updater as binary_updater
+import octobot.updater.python_updater as python_updater
+import octobot.updater.docker_updater as docker_updater
+
+
+def create_updater():
```
Comment (GuillaumeDSM) [inline octobot/updater/updater_factory.py:25]: 👍

File: octobot/updater/binary_updater.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        release_asset_name, matching_asset_binary = await self._get_asset_from_release_data()
+        new_binary_file = f"{release_asset_name}{self.NEW_BINARY_SUFFIX}"
+        if matching_asset_binary is None:
+            self.logger.error(f"Error when downloading latest version binary : Release not found on server")
+            return None
+        async with aiofiles.open(new_binary_file, 'wb+') as downloaded_file:
+            await aiohttp_util.download_stream_file(output_file=downloaded_file,
+                                                    file_url=matching_asset_binary["browser_download_url"],
+                                                    aiohttp_session=aiohttp.ClientSession(),
+                                                    is_aiofiles_output_file=True)
```
Comment (GuillaumeDSM) [inline octobot/updater/binary_updater.py:134]: 👍

File: octobot/updater/binary_updater.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                       f"{DeliveryArchitectureName.ARM_X64.value}{DeliveryPlatformsExtension.LINUX.value}"
+        return None
+
+    async def _download_binary(self):
+        release_asset_name, matching_asset_binary = await self._get_asset_from_release_data()
+        new_binary_file = f"{release_asset_name}{self.NEW_BINARY_SUFFIX}"
+        if matching_asset_binary is None:
+            self.logger.error(f"Error when downloading latest version binary : Release not found on server")
+            return None
+        async with aiofiles.open(new_binary_file, 'wb+') as downloaded_file:
```
Comment (GuillaumeDSM) [inline octobot/updater/binary_updater.py:130]: I think we should log something before starting the download to tell what is going on

**Review by GuillaumeDSM — APPROVED** [review]
🎉

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -128,18 +128,18 @@ jobs:
       - name: run tests in docker
         if: github.event_name != 'push'
         run: |
-          docker run -i -e TENTACLES_REPOSITORY=dev-tentacles -e TENTACLES_URL_TAG=${GITHUB_HEAD_REF////_} ${{ env.IMAGE }}:${{ env.TEST }}
+          docker run -i -e TENTACLES_REPOSITORY=dev-tentacles -e TENTACLES_URL_TAG=${GITHUB_HEAD_REF////_} -e DISABLE_SENTRY=True ${{ env.IMAGE }}:${{ env.TEST }}
```
Comment (GuillaumeDSM) [inline .github/workflows/docker.yml:131]: no sentry in docker ?

Comment (Herklos) [reply .github/workflows/docker.yml:131]: Yes when testing

File: octobot/updater/python_updater.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                if resp.status == 200:
+                    return json.loads(text)
+            return None
+        except Exception as e:
+            self.logger.debug(f"Error when fetching latest pypi package data : {e}")
+
+    def _get_latest_pypi_version_from_data(self, pypi_package_data):
+        try:
+            releases = pypi_package_data.get('releases', [])
+            version = parse('0')
```
Comment (GuillaumeDSM) [inline octobot/updater/python_updater.py:61]: Not sure I understand the purpose of this line

File: octobot/updater/python_updater.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                if not ver.is_prerelease:
+                    version = max(version, ver)
+            return version
+        except Exception as e:
+            self.logger.debug(f"Failed to parse pypi package version : {e}")
+            return None
+
+    async def update_impl(self) -> bool:
+        try:
+            if self.use_git:
```
Comment (GuillaumeDSM) [inline octobot/updater/python_updater.py:73]: 👍

**Review by GuillaumeDSM — APPROVED** [review]
👍

File: octobot/updater/python_updater.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+class PythonUpdater(updater_class.Updater):
+    def __init__(self):
+        super().__init__()
+        self.use_git = os.path.isdir(".git")
+
+    async def get_latest_version(self):
+        if self.use_git:
+            try:
+                self._run_git_fetch()
```
Comment (GuillaumeDSM) [inline octobot/updater/python_updater.py:37]: should we quickly check if the user has git and if not log an error message saying git is required with a link to download it ?

File: octobot/updater/python_updater.py (hunk 5 above)
Comment (Herklos) [reply octobot/updater/python_updater.py:61]: This is the initial version, used to be compared with max()

File: octobot/updater/python_updater.py (hunk 7 above)
Comment (Herklos) [reply octobot/updater/python_updater.py:37]: You mean the user has a ".git" folder without having git installed on his computer ?

File: octobot/updater/python_updater.py (hunk 5 above)
Comment (GuillaumeDSM) [reply octobot/updater/python_updater.py:61]: Oh, right !

File: octobot/updater/python_updater.py (hunk 7 above)
Comment (GuillaumeDSM) [reply octobot/updater/python_updater.py:37]: Right, I guess we can try without check

File: octobot/cli.py [hunk 8]
```diff
@@ -226,6 +229,8 @@ def octobot_parser(parser):
                         action='store_true')
     parser.add_argument('-s', '--simulate', help='Force OctoBot to start with the trader simulator only.',
                         action='store_true')
+    parser.add_argument('-u', '--update', help='Update OctoBot to latest version.',
```
Comment (GuillaumeDSM) [inline octobot/cli.py:232]: 👍

File: octobot/cli.py [hunk 9]
```diff
@@ -162,6 +162,9 @@ def start_octobot(args):
             # set community identifier
             bot.community_auth.identifier = args.identifier[0]
 
+        if args.update:
+            return commands.update_bot(bot.octobot_api)
```
Comment (GuillaumeDSM) [inline octobot/cli.py:166]: 👍

**Review by GuillaumeDSM — APPROVED** [review]
🎉

**Pull request merged**